pkg/lockless_circular_buffer/circular_buffer.go

```diff
@@ -0,0 +1,331 @@
+		}
+	}
+
+	return false
+}
+
+func (cb *CircularBuffer[T]) Pop() (T, bool) {
+	var zero T
+	const maxAttempts = 5
+
+	for i := 0; i < maxAttempts; i++ {
+		head := cb.head.Load()
+		tail := cb.tail.Load()
+
+		if head == tail {
+			return zero, false
+		}
+
+		nextHead := (head + 1) & cb.mask
+		if cb.head.CompareAndSwap(head, nextHead) {
+			item := cb.buffer[head]
+			return item, true
+		}
+
+		if i > 1 {
+			runtime.Gosched()
+		}
+	}
+	return zero, false
+}
+
+func (cb *CircularBuffer[T]) PushBatch(items []T) int {
+	if len(items) == 0 {
+		return 0
+	}
+
+	head := cb.head.Load()
+	tail := cb.tail.Load()
+
+	var availableSpace uint32
+	if head <= tail {
+		availableSpace = cb.size - (tail - head) - 1
+	} else {
+		availableSpace = head - tail - 1
+	}
+
+	batchSize := uint32(len(items))
+	if batchSize > availableSpace {
+		batchSize = availableSpace
+	}
+
+	if batchSize == 0 {
+		return 0
+	}
+
+	pushed := uint32(0)
+	for pushed < batchSize {
+		tail = cb.tail.Load()
+		head = cb.head.Load()
+
+		if head <= tail {
+			availableSpace = cb.size - (tail - head) - 1
+		} else {
+			availableSpace = head - tail - 1
+		}
+
+		if availableSpace == 0 {
+			break
+		}
+
+		currentBatchSize := batchSize - pushed
+		if currentBatchSize > availableSpace {
+			currentBatchSize = availableSpace
+		}
+
+		newTail := (tail + currentBatchSize) & cb.mask
+
+		if cb.tail.CompareAndSwap(tail, newTail) {
+			for i := uint32(0); i < currentBatchSize; i++ {
+				slotIndex := (tail + i) & cb.mask
+				cb.buffer[slotIndex] = items[pushed+i]
+			}
+			pushed += currentBatchSize
+		}
+	}
+
+	return int(pushed)
+}
+
+func (cb *CircularBuffer[T]) PushBatchBlocking(items []T) {
+	remaining := items
+	for len(remaining) > 0 {
+		pushed := cb.PushBatch(remaining)
+		if pushed == 0 {
+			runtime.Gosched()
+			continue
+		}
+		remaining = remaining[pushed:]
+	}
+}
+
+func (cb *CircularBuffer[T]) PopBatch(maxItems int) (int, []T) {
+	if maxItems <= 0 {
+		return 0, nil
+	}
+
+	head := cb.head.Load()
+	tail := cb.tail.Load()
+
+	var availableItems uint32
+	if tail >= head {
+		availableItems = tail - head
+	} else {
+		availableItems = cb.size - (head - tail)
+	}
+
+	batchSize := uint32(maxItems)
+	if batchSize > availableItems {
+		batchSize = availableItems
+	}
+
+	if batchSize == 0 {
+		return 0, nil
+	}
+
+	result := make([]T, 0, batchSize)
+	popped := uint32(0)
+
+	for popped < batchSize {
+		head = cb.head.Load()
+		tail = cb.tail.Load()
+
+		if tail >= head {
+			availableItems = tail - head
+		} else {
+			availableItems = cb.size - (head - tail)
+		}
+
+		if availableItems == 0 {
+			break
+		}
+
+		currentBatchSize := batchSize - popped
+		if currentBatchSize > availableItems {
+			currentBatchSize = availableItems
+		}
+
+		newHead := (head + currentBatchSize) & cb.mask
+
+		if cb.head.CompareAndSwap(head, newHead) {
+			for i := uint32(0); i < currentBatchSize; i++ {
+				slotIndex := (head + i) & cb.mask
+				result = append(result, cb.buffer[slotIndex])
+			}
```

Comment:
this can be implemented in at most 2 copies? after head has been updated is the read range safe from concurrent writes?